src/app/components/ManualEntry.svelte

```diff
@@ -1,9 +1,80 @@
 <script lang="ts">
-  /**
-   * Handles the click when the delete button is pressed
-   * TODO
-   */
-  const deleteRow = (): void => {};
+  import Button from "./Button.svelte";
+  import worker from "../spawn-worker";
+
+  interface DictionaryEntry {
+    word: string;
+    count?: number;
+  }
+
+  export let tableData: {
+    name: string;
+    data: DictionaryEntry[];
+  };
+
+  let columns = [
+    {
+      label: "Word",
+      name: "word",
+    },
+    {
+      label: "Count",
+      name: "count",
+    },
+    {
+      label: "Action",
+      name: "action",
+    },
+  ];
+  $: rowDataFromManualEntry = tableData.data;
+  $: validDictionary =
+    validateRowsData(rowDataFromManualEntry) &&
+    validateTableName(tableData.name);
+  $: if (validDictionary) {
+    worker.addManualEntryDictionaryToProject(tableData);
+  }
+
+  const validateTableName = (tableName: string): boolean => {
+    return (
+      tableName !== "" &&
+      tableName !== undefined &&
+      tableName !== null &&
+      !tableName.match(/^([\s\t\r\n]*)$/)
+    );
+  };
+
+  const validateRowsData = (rowsData: DictionaryEntry[]): boolean => {
+    return rowsData.every((rowData) => {
+      const word = rowData.word;
+      return (
+        word !== "" &&
+        word !== undefined &&
+        word !== null &&
+        !word.match(/^([\s\t\r\n]*)$/)
+      );
+    });
+  };
+
+  const deleteRow = (i: number): void => {
+    rowDataFromManualEntry.splice(i, 1);
+    rowDataFromManualEntry = rowDataFromManualEntry;
+  };
+
+  const addNewRow = (): void => {
+    rowDataFromManualEntry.push({ word: "" });
+    rowDataFromManualEntry = rowDataFromManualEntry;
+  };
+
+  const saveTableData = async () => {
+    const validDictionary: boolean =
+      validateRowsData(rowDataFromManualEntry) &&
+      validateTableName(tableData.name);
+    if (validDictionary) {
+      const numOfWordStored = await worker.addManualEntryDictionaryToProject(
+        tableData
+      );
+    }
```

Comment:
There's some duplicated code here with the reactive statementsthat handle autosave (`$: if (validDictionary) ...`). Can you reduce the duplication here?

Reply:
Sure, I will try to factor out the duplication 😅